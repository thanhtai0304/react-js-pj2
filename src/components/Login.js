import React, { useState } from "react";
import {
  Segment,
  Header,
  Grid,
  Image,
  Form,
} from "semantic-ui-react";
import "semantic-ui-css/semantic.min.css";
import { connect } from "react-redux";
import { setAuthUser } from "../store/actions/authUser";
import { withRouter } from "react-router-dom";
import GridContainer from "./GridContainer";

const Login = ({ users, setAuthUser, history, location }) => {
  const [userId, setUserId] = useState("");
  const [disabled, setDisabled] = useState(true);

  const onSubmit = (event) => {
    event.preventDefault();

    setAuthUser(userId);
    history.push(location.pathname);
  };

  const onChange = (e, { value }) => {
    setUserId(value);
    setDisabled(false);
  };

  return (
    <React.Fragment>
      <GridContainer>
        <Segment.Group>
          <Header as="h4" block attached="top" textAlign="center">
            <Header.Content>
              Animal Pools
            </Header.Content>
            <Header.Subheader>Please sign in to continue</Header.Subheader>
          </Header>
          <Grid padded textAlign="center">
            <Grid.Row>
              <Grid.Column width={16}>
                <Image src="/images/logo.png" size="medium" centered />
                <br />
                <Form onSubmit={onSubmit}>
                  <Header as="h2" color="blue">
                    Sign In
                  </Header>
                  <Form.Dropdown
                    placeholder="Select a animal"
                    fluid
                    selection
                    scrolling
                    value={userId}
                    required
                    onChange={onChange}
                    options={
                      (users &&
                        users.map((user) => ({
                          key: user.id,
                          text: user.name,
                          value: user.id,
                          image: { avatar: true, src: user.avatarURL },
                        }))) ||
                      []
                    }
                  />
                  <Form.Button
                    content="Login"
                    primary
                    fluid
                    disabled={disabled}
                  />
                </Form>
              </Grid.Column>
            </Grid.Row>
          </Grid>
        </Segment.Group>
      </GridContainer>
    </React.Fragment>
  );
};

const mapStateToProps = (state) => {
  return {
    users: state.users.users,
    authUser: state.authUser && state.authUser.user,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setAuthUser: (userId) => dispatch(setAuthUser(userId)),
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(Login));
