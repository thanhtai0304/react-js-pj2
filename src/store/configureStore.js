import { createStore, combineReducers, applyMiddleware } from "redux";
import users from "./reducers/users";
import questions from './reducers/questions' 
import thunk from 'redux-thunk';
import authUser  from "./reducers/authUser";

const rootReducer = combineReducers({
  users: users,
  questions: questions,
  authUser: authUser,
});

const store = createStore(rootReducer, applyMiddleware(thunk));

// Exports
export { store };
