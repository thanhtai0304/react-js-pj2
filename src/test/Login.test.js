import { fireEvent, render, screen } from '@testing-library/react';
import Login from '../components/Login';
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { store } from "../store/configureStore";

describe('Login', () => {
  it('Should be rendered successfully', () => {
    const renderedComponent = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    )
    expect(renderedComponent).toBeDefined();
    expect(renderedComponent).toMatchSnapshot()
  })

  it('should disable on submit button', () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    expect(component).toBeDefined();

    const submitButtonElement = component.getByTestId("submit");
    expect(submitButtonElement).toBeInTheDocument();

    fireEvent.click(submitButtonElement);
    expect(submitButtonElement).toBeDisabled();

  })

  it('should login after click submit button', () => {
    const users = [
      { id: 1, name: 'User 1', avatarURL: 'url1' },
      { id: 2, name: 'User 2', avatarURL: 'url2' },
    ];

    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login/>
        </BrowserRouter>
      </Provider>
    );

    expect(component).toBeDefined();

    const dropdown = component.getByTestId('userId');
    expect(dropdown).toBeDefined();
    fireEvent.change(dropdown,{ target: { value: 'user1' } });
    expect(dropdown.value).toBe('2');
    // fireEvent.change(dropdown, { target: { value: '2' } });
    // fireEvent.change(usernameInputElement, {target: {value: 'rhinoceros'}});
    // expect(usernameInputElement.value).toBe("");

  })

})
