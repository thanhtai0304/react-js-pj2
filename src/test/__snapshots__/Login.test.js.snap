// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Should initialize the login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui centered vertically padded one column grid"
      >
        <div
          class="row"
        >
          <div
            class="column"
            style="max-width: 1100px;"
          >
            <div
              class="ui segments"
            >
              <h4
                class="ui block top attached center aligned header"
              >
                <div
                  class="content"
                >
                  Animal Pools
                </div>
                <div
                  class="sub header"
                  data-testid="login-heading"
                >
                  Please sign in to continue
                </div>
              </h4>
              <div
                class="ui padded center aligned grid"
              >
                <div
                  class="row"
                >
                  <div
                    class="sixteen wide column"
                  >
                    <img
                      class="ui medium centered image"
                      src="/images/logo.png"
                    />
                    <br />
                    <form
                      class="ui form"
                    >
                      <h2
                        class="ui blue header"
                      >
                        Sign In
                      </h2>
                      <div
                        class="required field"
                      >
                        <div
                          aria-expanded="false"
                          class="ui fluid selection scrolling dropdown"
                          data-testid="userId"
                          required=""
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-atomic="true"
                            aria-live="polite"
                            class="default text"
                            role="alert"
                          >
                            Select a animal
                          </div>
                          <i
                            aria-hidden="true"
                            class="dropdown icon"
                          />
                          <div
                            class="menu transition"
                          />
                        </div>
                      </div>
                      <div
                        class="disabled field"
                      >
                        <button
                          class="ui fluid primary disabled button"
                          data-testid="submit"
                          disabled=""
                          tabindex="-1"
                        >
                          Login
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui centered vertically padded one column grid"
    >
      <div
        class="row"
      >
        <div
          class="column"
          style="max-width: 1100px;"
        >
          <div
            class="ui segments"
          >
            <h4
              class="ui block top attached center aligned header"
            >
              <div
                class="content"
              >
                Animal Pools
              </div>
              <div
                class="sub header"
                data-testid="login-heading"
              >
                Please sign in to continue
              </div>
            </h4>
            <div
              class="ui padded center aligned grid"
            >
              <div
                class="row"
              >
                <div
                  class="sixteen wide column"
                >
                  <img
                    class="ui medium centered image"
                    src="/images/logo.png"
                  />
                  <br />
                  <form
                    class="ui form"
                  >
                    <h2
                      class="ui blue header"
                    >
                      Sign In
                    </h2>
                    <div
                      class="required field"
                    >
                      <div
                        aria-expanded="false"
                        class="ui fluid selection scrolling dropdown"
                        data-testid="userId"
                        required=""
                        role="listbox"
                        tabindex="0"
                      >
                        <div
                          aria-atomic="true"
                          aria-live="polite"
                          class="default text"
                          role="alert"
                        >
                          Select a animal
                        </div>
                        <i
                          aria-hidden="true"
                          class="dropdown icon"
                        />
                        <div
                          class="menu transition"
                        />
                      </div>
                    </div>
                    <div
                      class="disabled field"
                    >
                      <button
                        class="ui fluid primary disabled button"
                        data-testid="submit"
                        disabled=""
                        tabindex="-1"
                      >
                        Login
                      </button>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login Should be rendered successfully 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui centered vertically padded one column grid"
      >
        <div
          class="row"
        >
          <div
            class="column"
            style="max-width: 1100px;"
          >
            <div
              class="ui segments"
            >
              <h4
                class="ui block top attached center aligned header"
              >
                <div
                  class="content"
                >
                  Animal Pools
                </div>
                <div
                  class="sub header"
                  data-testid="login-heading"
                >
                  Please sign in to continue
                </div>
              </h4>
              <div
                class="ui padded center aligned grid"
              >
                <div
                  class="row"
                >
                  <div
                    class="sixteen wide column"
                  >
                    <img
                      class="ui medium centered image"
                      src="/images/logo.png"
                    />
                    <br />
                    <form
                      class="ui form"
                    >
                      <h2
                        class="ui blue header"
                      >
                        Sign In
                      </h2>
                      <div
                        class="required field"
                      >
                        <div
                          aria-expanded="false"
                          class="ui fluid selection scrolling dropdown"
                          data-testid="userId"
                          required=""
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-atomic="true"
                            aria-live="polite"
                            class="default text"
                            role="alert"
                          >
                            Select a animal
                          </div>
                          <i
                            aria-hidden="true"
                            class="dropdown icon"
                          />
                          <div
                            class="menu transition"
                          />
                        </div>
                      </div>
                      <div
                        class="disabled field"
                      >
                        <button
                          class="ui fluid primary disabled button"
                          data-testid="submit"
                          disabled=""
                          tabindex="-1"
                        >
                          Login
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui centered vertically padded one column grid"
    >
      <div
        class="row"
      >
        <div
          class="column"
          style="max-width: 1100px;"
        >
          <div
            class="ui segments"
          >
            <h4
              class="ui block top attached center aligned header"
            >
              <div
                class="content"
              >
                Animal Pools
              </div>
              <div
                class="sub header"
                data-testid="login-heading"
              >
                Please sign in to continue
              </div>
            </h4>
            <div
              class="ui padded center aligned grid"
            >
              <div
                class="row"
              >
                <div
                  class="sixteen wide column"
                >
                  <img
                    class="ui medium centered image"
                    src="/images/logo.png"
                  />
                  <br />
                  <form
                    class="ui form"
                  >
                    <h2
                      class="ui blue header"
                    >
                      Sign In
                    </h2>
                    <div
                      class="required field"
                    >
                      <div
                        aria-expanded="false"
                        class="ui fluid selection scrolling dropdown"
                        data-testid="userId"
                        required=""
                        role="listbox"
                        tabindex="0"
                      >
                        <div
                          aria-atomic="true"
                          aria-live="polite"
                          class="default text"
                          role="alert"
                        >
                          Select a animal
                        </div>
                        <i
                          aria-hidden="true"
                          class="dropdown icon"
                        />
                        <div
                          class="menu transition"
                        />
                      </div>
                    </div>
                    <div
                      class="disabled field"
                    >
                      <button
                        class="ui fluid primary disabled button"
                        data-testid="submit"
                        disabled=""
                        tabindex="-1"
                      >
                        Login
                      </button>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Login Should initialize the login 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui centered vertically padded one column grid"
      >
        <div
          class="row"
        >
          <div
            class="column"
            style="max-width: 1100px;"
          >
            <div
              class="ui segments"
            >
              <h4
                class="ui block top attached center aligned header"
              >
                <div
                  class="content"
                >
                  Animal Pools
                </div>
                <div
                  class="sub header"
                  data-testid="login-heading"
                >
                  Please sign in to continue
                </div>
              </h4>
              <div
                class="ui padded center aligned grid"
              >
                <div
                  class="row"
                >
                  <div
                    class="sixteen wide column"
                  >
                    <img
                      class="ui medium centered image"
                      src="/images/logo.png"
                    />
                    <br />
                    <form
                      class="ui form"
                    >
                      <h2
                        class="ui blue header"
                      >
                        Sign In
                      </h2>
                      <div
                        class="required field"
                      >
                        <div
                          aria-expanded="false"
                          class="ui fluid selection scrolling dropdown"
                          data-testid="userId"
                          required=""
                          role="listbox"
                          tabindex="0"
                        >
                          <div
                            aria-atomic="true"
                            aria-live="polite"
                            class="default text"
                            role="alert"
                          >
                            Select a animal
                          </div>
                          <i
                            aria-hidden="true"
                            class="dropdown icon"
                          />
                          <div
                            class="menu transition"
                          />
                        </div>
                      </div>
                      <div
                        class="disabled field"
                      >
                        <button
                          class="ui fluid primary disabled button"
                          data-testid="submit"
                          disabled=""
                          tabindex="-1"
                        >
                          Login
                        </button>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui centered vertically padded one column grid"
    >
      <div
        class="row"
      >
        <div
          class="column"
          style="max-width: 1100px;"
        >
          <div
            class="ui segments"
          >
            <h4
              class="ui block top attached center aligned header"
            >
              <div
                class="content"
              >
                Animal Pools
              </div>
              <div
                class="sub header"
                data-testid="login-heading"
              >
                Please sign in to continue
              </div>
            </h4>
            <div
              class="ui padded center aligned grid"
            >
              <div
                class="row"
              >
                <div
                  class="sixteen wide column"
                >
                  <img
                    class="ui medium centered image"
                    src="/images/logo.png"
                  />
                  <br />
                  <form
                    class="ui form"
                  >
                    <h2
                      class="ui blue header"
                    >
                      Sign In
                    </h2>
                    <div
                      class="required field"
                    >
                      <div
                        aria-expanded="false"
                        class="ui fluid selection scrolling dropdown"
                        data-testid="userId"
                        required=""
                        role="listbox"
                        tabindex="0"
                      >
                        <div
                          aria-atomic="true"
                          aria-live="polite"
                          class="default text"
                          role="alert"
                        >
                          Select a animal
                        </div>
                        <i
                          aria-hidden="true"
                          class="dropdown icon"
                        />
                        <div
                          class="menu transition"
                        />
                      </div>
                    </div>
                    <div
                      class="disabled field"
                    >
                      <button
                        class="ui fluid primary disabled button"
                        data-testid="submit"
                        disabled=""
                        tabindex="-1"
                      >
                        Login
                      </button>
                    </div>
                  </form>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
